### ISSUE TITLE
Is there authentication guide to Trains

### ISSUE THREAD

**Issue by shlomiken** [allegroai/trains#13]
Hi
i used the AMI to deplot to AWS. but i can see Everyone can login .
is there a guide for authentication ?

**Comment by bmartinn** [comment]
See matching thread here: https://github.com/allegroai/trains-server/issues/3
Feel free to vote with reaction :+1: 

TL;DL
*TRAINS-server* is for internal use only, the main goal is full transparency, so everyone in the company can login (which is mostly for read access).

If you are using the AMI distribution, I *highly* recommend configuring the security-group of the instance to limit the access to the server for internal company IP group...

**Comment by bmartinn** [comment]
> in the AMI deployment , where exactly is the trains-init script , could not find it , not even in the docker instances (webserver/api)

The `trains-init` script is just an easier way to configure credentials for the SDK (i.e. the TRAINS python package). The *TRAINS-server* does not need any credentials configured.

Once you login to the web-app (with any username, as currently there is not limitation)
You can generate your key/secret credentials pair (which the TRAINS package is using, in a similar way to AWS credentials etc) in the profile page (which is currently called 'admin' but will be changed in the next release) at http://localhost:8080/admin

**Comment by shlomiken** [comment]
thanks for the explanations.
regarding limiting IP addresses - this is a problem , as we use sagemaker to train models.
and i cannot tell which IP address sagemaker will use. (in order to report the metrics you collect)

**Comment by shlomiken** [comment]
@bmartinn - i have put the instance in AWS behind a LB , so it could be available to our Sagemaker server 
but i get 405 method not allowed on POST - i believe this is because of there headers sent to my browser
if i go directly to the instance i get these headers which suggest that the local IP address is used as allowed origin
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: http://10.121.1.213:8080
meaning this cannot be put behind ELB ?
thanks
Shlomi

**Comment by bmartinn** [comment]
The `demoapp.trainsai.io` is an almost exact replica of the pre-installed *trains-server* AMI and it works fine behind an ELB.

@shlomiken Can you please share the *trains-server* logs? (default location /opt/trains/logs)

**Comment by shlomiken** [comment]
[trains.tar.gz](https://github.com/allegroai/trains/files/3331668/trains.tar.gz)
attached
i believe this is CORS issue - as when i request directly from the instance i see OPTIONS is being sent to server.
i might be missing something in configuration that say hey this is my domain name for this server.

**Comment by bmartinn** [comment]
The webserver's CORS configuration uses origin "*", which means any request origin is allowed, and the response headers will simply echo the origin - this is the reason you see `Access-Control-Allow-Origin: https://demoapp.trainsai.io` in our demo server's response, and `Access-Control-Allow-Origin: http://10.121.1.213:8080` in yours (when you do succeed reaching the server).

Just to make sure, the responses we're talking about are those received by your browser when trying to access the webserver in port 8080, right?

This issue might be related to a HTTPS redirect issue, can you please describe how you configured the ELB?

**Comment by shlomiken** [comment]
Hi @bmartinn - the ELB is configured with both 8008, 80 and 443 listeners. that forward to the instance on port 8080 or 8008.
this is an Application LB . 
Is there special rule i need to add ? can you share you configuration ?
thanks
Shlomi

**Comment by bmartinn** [comment]
I've just set up an AWS EC2 instance using the **trains-server** AMI to try and replicate your use-case. It works as expected, and my browser can successfully access the webserver with the URL `http://<my-elb>.us-east-1.elb.amazonaws.com:8080`.

I've used the following ELB configuration:
- ELB Type: HTTP/HTTPS load balancer.
- Listeners:
  * HTTP:8008 listener, forwarded to a target group:
    * Target type: Instance
    * Protocol: HTTP
    * Port 8008
  * HTTP:8080 listener, forwarded to a target group:
    * Target type: Instance
    * Protocol: HTTP
    * Port 8080

**Comment by shlomiken** [comment]
the difference is i use port 80 to mask port 8080 cause our sec groups do not allow 8080 traffic ( just like your demo server)
can you please share the configuration of the demo server ?
maybe its conf files without tokens?

**Comment by bmartinn** [comment]
You're correct, I've reproduced the issue right now and the reason is the different port used for accessing the webserver. The webapp assumes the webserver uses port 8080. Since in your case the webserver uses port 80, the webapp fails to access the apiserver.

I'll add this issue as a configuration feature request.

The demo server uses subdomains (demoapp, demoapi) instead of port numbers to redirect traffic to the webserver and the apiserver (the webapp can detect these names and adjust accordingly).

Since in your case, sagemaker only requires access to port 8008 (which is the port used by the **trains** package), you can:
- Use ELB with a single 8008 listener and configure sagemaker/trains to use the external ELB address with port 8008
- Use the EC2 instance private IP port 8080 in your browser, like you currently do

**Comment by shlomiken** [comment]
@bmartinn - how can i use subdomains as well ?

**Comment by bmartinn** [comment]
You can configure ELB to direct the **app** subdomain to port 8080 and the **api** subdomain to port 8008. Please make sure to use HTTP and not HTTPS as changing this will require reconfiguration of the webserver (which is part of the pre-built AMI).

**Comment by shlomiken** [comment]
@bmartinn sure , but how the UI knows to which api server to address. 
my problem is the the UI is also not functioning cause of this issue - we have no problem with sage maker calls.

**Comment by bmartinn** [comment]
The UI can detect the **app** subdomain and use the **api** subdomain to access the server.

**Comment by shlomiken** [comment]
i have created 2 separated domains in AWS.
![image](https://user-images.githubusercontent.com/32815285/60282553-eeb66200-990f-11e9-9f50-60590fa4db78.png)
i have tried that (although sounds weird) - but the api requests still try to get to trainsapp

**Comment by bmartinn** [comment]
Use **app** and **api**, we in `app.trains.kenshoo-lab.com`and `api.trains.kenshoo-lab.com`.

**Comment by bmartinn** [comment]
Apologies, it seems the code related to custom subdomains redirection is not part of the docker build. I'll include it in the next release (soon). For now, you can use the ports approach discussed earlier.
